consensus/types.go

```diff
@@ -1,53 +1,168 @@
+	// We are able to simplify the GetQC() to this behavior because this QC is only needed to construct
+	// a timeout QC for the next block in the event of a timeout. So, this QC will always be the latest QC
+	// at the current chain's tip that subsequent blocks will build on top of.
+	GetQC() QuorumCertificate
+}
+
+type QuorumCertificate interface {
+	GetBlockHash() BlockHash
+	GetView() uint64
+	GetSignersList() *bitset.Bitset
+	GetAggregatedSignature() *bls.Signature
+}
+
+// An instance of FastHotStuffConsensus is a self-contained module that represents a single node running the
+// Fast HotStuff consensus protocol. The module is initialized at the current chain's tip, with a given
+// block hash, block height, view number, and validator set. The module is simplified and does not know
+// whether its role is that of a block proposer or a validator.
+//
+// Given a block that's at the tip of the current chain, this module maintains its own internal data structures
+// and runs internal timers that handles all of the following:
+//   - Tracking of the current view, incrementing the view during timeouts, and computing exponential
+//     back-off durations during consecutive timeouts
+//   - Aggregation of votes and QC construction for the current block
+//   - Aggregation of timeout messages for the current view
+//   - Signaling its caller when it can vote on the current chain tip
+//   - Signaling its caller when it has timed out the current view
+//   - Signaling its caller when it has a QC for the current block
+//   - Signaling its caller when it has a timeout QC for the current view
+//
+// When a new block is connected to the chain, the caller is expected to update the chain tip. The module
+// resets all internal data structures and timers to handle all of the above based on the new chain tip.
+//
+// This module is very simply and only houses the logic that decide what action to perform next given the
+// current chain tip. The module does not track the history of blocks, and instead needs its caller to
+// update the block at the current chain tip. It expects its caller to maintain the block chain,
+// the index of all past blocks, to perform QC validations for incoming blocks, to handle the commit rule,
+// and only then to pass the validated chain tip. Note: this module takes the provided chain tip as a
+// trusted input and does NOT validate any incoming blocks. This also mean the module expects its caller to
+// track historical vote and timeout messages it has sent so as to not vote more than once at a given view
+// or block height.
 type FastHotStuffConsensus struct {
-	nextBlockProposalTime time.Time
-	nextTimeoutTime       time.Time
+	lock sync.RWMutex
+
+	blockConstructionCadence time.Duration
+	timeoutBaseDuration      time.Duration
 
-	internalTimersUpdated chan interface{}
-	quit                  chan interface{}
+	nextBlockConstructionTime time.Time
+	nextTimeoutTime           time.Time
```

Comment:
should we name these to be `nextBlockConstructionTimestamp` and `nextTimeoutTimestamp`? `Timestamp` seems a little less ambiguous than `Time`

Reply:
Sure